README.md
# Arduino Core for Arancino SAMD21 Boards
![Release](https://img.shields.io/github/v/release/smartmeio/arancino-core-samd21?style=plastic)

This core is a fork of the [Arduino SAMD21](https://github.com/arduino/ArduinoCore-samd) core modified to support the functionality of the Arancino architecture.

# Supported Boards
* Arancino
* Arancino Mignon

## Installation on Arduino IDE

This core is available as a package in the Arduino IDE cores manager. If you want to install it:

  1. Open the **Preferences** of the Arduino IDE.
  2. Add this URL `https://raw.githubusercontent.com/smartmeio/arancino-boards/master/package_smartmeio_index.json` in the **Additional Boards Manager URLs** field, and click OK.
  3. Open the **Boards Manager** (menu `Tools` -> `Board` -> `Board Manager...`)
  4. Install **Arancino SAMD Boards**
  5. Select one of the boards under **Arancino SAMD Boards** in `Tools` -> `Board` menu

## Installation on PlatformIO and Visual Studio Code
To create a project with Visual Studio Code and PlatformIO it is necessary to initially create a project for `Arduino M0` and then modify the `platformio.ini` file in order to point to the Arancino packages. The `platformio.ini` file must be modififed in order to contains this configuration:
```
[env:arancinosamd]
platform = https://github.com/smartmeio/platform-atmelsam.git#8.1.0-arancino
board = arancino
framework = arduino
upload_protocol = sam-ba
platform_packages = smartme-io/framework-arduino-samd-arancino@https://git.smartme.io/smartme.io/arancino/ide/smartmeio-arancino-platform.git
lib_deps = https://github.com/smartmeio/arancino-library
  https://github.com/smartmeio/Arduino-FreeRTOS-SAMD21
upload_port = ...
```
Any other used library must be included as dependency through `lib_deps` (as for the Arancino library) or saved under the `lib` folder; in the latter case the project structure should look like:
```
include
lib
+-- your_library
     +-- examples
     +-- include
     +-- keywords.txt
     +-- library.json
     +-- src
src
+-- main.cpp
test
```

Alternatively, you can set an extra directory in which to search for libraries. For example, if you want to include all the libraries installed in the Arduino IDE, you can specify it adding this lines to the `platformio.ini` configuration file:
```
lib_extra_dirs = ~/Arduino/libraries
```

# Uploading Sketches

### Arduino IDE
Select the appropriate serial port shown in the Arduino `Tools` -> `Port` -> `Serial Port` menu once (this setting will stick and does not need to be
touched for multiple uploads). This selection allows the auto-reset tool to identify the proper device to reset.

### PlatformIO
Set the appropriate serial port into the `platformio.ini` file, e.g.:
```
upload_port = /dev/ttyACM1
```
Then hit the upload button and your sketch should upload and run.

In some cases the MCU will encounter a hard hang and its USB port will not respond to the auto-reset request. Should this happen, put the MCU in `bootloader mode` double-tapping the reset button and then retry the upload on the new serial port.

# Debugging with OpenOCD and GDB

### PlatformIO
To debug the code you need an external debugger (e.g. Atmel EDBG, Ateml ICE, etc).

This additional debugging information can be added to the `platformio.ini` file in order to improve the debugging process.
```
build_type = debug
lib_archive = no
```

# Enabling FreeRTOS support
This core supports FreeRTOS: to include and enable it just include the `Arancino.h` library and, depending on the used IDE, enable it. 

### Arduino IDE
Through the menu `Tools` -> `Using FreeRTOS` -> `Yes`.

### PlatformIO
Add an extra flag to the `platformio.ini` configuration file:
```
build_flags = -DUSEFREERTOS -DENABLE_CALLOC_REALLOC -Wl,--wrap=malloc -Wl,--wrap=free -Wl,--wrap=calloc -Wl,--wrap=realloc
```


# Licensing and Credits
* The [Arduino SAMD21 core](https://github.com/arduino/ArduinoCore-samd) is developed and maintained by Arduino LLC in collaboration with Atmel.
* The [Arduino IDE and ArduinoCore-API](https://arduino.cc) are developed and maintained by the Arduino team. The IDE is licensed under GPL.
* [FreeRTOS](https://freertos.org) is Copyright Amazon.com, Inc. or its affiliates, and distributed under the MIT license.

This core has been developed by Arduino LLC in collaboration with Atmel, 
and customized by smartme.IO s.r.l.

```
  Copyright (c) 2019 smartme.IO.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
```

Arduino and Atmel are trademarks of their respective owners.


CHANGELOG
#### v 1.4.0 - 2023.04.06
* SerialUSB Thread-Safe [#863fztk29](https://app.clickup.com/t/863fztk29)
* Add StreamUtils Library [#863gant3w](https://app.clickup.com/t/863gant3w)
* Fix incompatibility between macros min, max and other [#863gantae](https://app.clickup.com/t/863gantae)

#### v 1.3.1 - 2022.09.27
* Added menu entry for FreeRTOS enabling [#30cgvhy](https://app.clickup.com/t/30cgvhy)
* Enabled heap functions wrapping [#30cgw4z] (https://app.clickup.com/t/30cgw4z)
* Rename build board for Arancino and Arancino Mignon[#2webywq](https://app.clickup.com/t/2webywq)

#### v 1.3.0 - 2021.11.16

* added FreeRTOS-SAMD21 library tag 2.3.0 that include freeRtos core version 10.2.1 [#nt82u1](https://app.clickup.com/t/nt82u1)
* enabled atomic mode for analogRead and analogWrite function [#kd95qa](https://app.clickup.com/t/kd95qa)
* definition of a new variable `MCU_FAMILY`[#1nf8f15](#https://app.clickup.com/t/1nf8f15)
* moved the libraries used by arancino-library into the core

#### v 1.2.2 - 2021.03.11

* main features of previous release
* ArduinoOTA tool updated to release v1.0.10 [#et1342](https://app.clickup.com/t/et1342)
* Dual Compatibility with PlatformIO (platform and tools) [#agym78](https://app.clickup.com/t/agym78)
* now the CI/CD script is only triggered by creation of tags

#### v 1.2.1 - 2021.01.14

* Dropping `build.memory_wrapping_flags` entry which is not anymore supported by FreeRTOS_samd library

#### v 1.2.0 - 2020.10.13

* Based on official Arduino samd core v1.8.9.
* Now ArancinoOTA is the default upload tool for Arancino board.
* Arancino Mignon can be programmed both with bossac (by serial port) or with ArancinoOTA (by OTA).
* Introduced a constant for Arancino Core version number: `ARANCINO_CORE_VERSION`.
* Introduced the possibility to program Arancino Mignon via RS485.
* Disabled file hex generation in build process to prevent and use only bin file to program mcu.
* Enabled PWM timer on gpio 2

#### v 1.1.0 - 2020.01.09

* Second release

#### v 1.0.0 - 2019.11.27

* First release


cores/arancino/Arduino.h
/*
  Arduino.h - Main include file for the Arduino SDK
  Copyright (c) 2014 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef Arduino_h
#define Arduino_h

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef bool boolean;
typedef uint8_t byte;
typedef uint16_t word;

// some libraries and sketches depend on this AVR stuff,
// assuming Arduino.h or WProgram.h automatically includes it...
//
#include "avr/pgmspace.h"
#include "avr/interrupt.h"
#include "avr/io.h"

#include "binary.h"
#include "itoa.h"

#ifdef __cplusplus
extern "C"{
#endif // __cplusplus

// Include Atmel headers
#include "sam.h"

#include "wiring_constants.h"

#define clockCyclesPerMicrosecond() ( SystemCoreClock / 1000000L )
#define clockCyclesToMicroseconds(a) ( ((a) * 1000L) / (SystemCoreClock / 1000L) )
#define microsecondsToClockCycles(a) ( (a) * (SystemCoreClock / 1000000L) )

void yield( void ) ;

/* system functions */
int main( void );
void init( void );

/* sketch */
void setup( void ) ;
void loop( void ) ;

#include "WVariant.h"

#ifdef __cplusplus
} // extern "C"
#endif

// The following headers are for C++ only compilation
#ifdef __cplusplus
  #include "WCharacter.h"
  #include "WString.h"
  #include "Tone.h"
  #include "WMath.h"
  #include "HardwareSerial.h"
  #include "pulse.h"
#endif
#include "delay.h"
#ifdef __cplusplus
  #include "Uart.h"
#endif

// Include board variant
#include "variant.h"

#include "wiring.h"
#include "wiring_digital.h"
#include "wiring_analog.h"
#include "wiring_shift.h"
#include "WInterrupts.h"

// undefine stdlib's abs if encountered
#ifdef abs
#undef abs
#endif // abs

#define min(a,b) ((a)<(b)?(a):(b))
#define max(a,b) ((a)>(b)?(a):(b))
#define abs(x) ((x)>0?(x):-(x))
#define constrain(amt,low,high) ((amt)<(low)?(low):((amt)>(high)?(high):(amt)))
#define round(x)     ((x)>=0?(long)((x)+0.5):(long)((x)-0.5))
#define radians(deg) ((deg)*DEG_TO_RAD)
#define degrees(rad) ((rad)*RAD_TO_DEG)
#define sq(x) ((x)*(x))

#define interrupts() __enable_irq()
#define noInterrupts() __disable_irq()

#define lowByte(w) ((uint8_t) ((w) & 0xff))
#define highByte(w) ((uint8_t) ((w) >> 8))

#define bitRead(value, bit) (((value) >> (bit)) & 0x01)
#define bitSet(value, bit) ((value) |= (1UL << (bit)))
#define bitClear(value, bit) ((value) &= ~(1UL << (bit)))
#define bitWrite(value, bit, bitvalue) (bitvalue ? bitSet(value, bit) : bitClear(value, bit))

#define bit(b) (1UL << (b))

#if (ARDUINO_SAMD_VARIANT_COMPLIANCE >= 10606)
// Interrupts
#define digitalPinToInterrupt(P)   ( P )
#endif

// Allows publishing the Beta core under samd-beta / arduino organization
#ifndef ARDUINO_ARCH_SAMD
#define ARDUINO_ARCH_SAMD
#endif

// Allows Arancino Lib to read this value and pass it to Arancino Module
// This value is modified in a CI/CD environment when a new tag is created
// This values are intended to be used by the Arancino library, if you want to use different values you can edit the ArancinoDefinitions.h file

#ifndef ARANCINO_CORE_VERSION
#define ARANCINO_CORE_VERSION "1.3.1"
#endif

#ifndef MCU_FAMILY
#define MCU_FAMILY "SAMD21"
#endif

#ifndef ARANCINO_SERIAL
#define ARANCINO_SERIAL SerialUSB
#endif

#ifndef ARANCINO_SERIAL_DEBUG
#define ARANCINO_SERIAL_DEBUG Serial
#endif

#ifndef ARANCINO_BAUDRATE
#define ARANCINO_BAUDRATE 4000000
#endif

#ifndef ARANCINO_TIMEOUT
#define ARANCINO_TIMEOUT 100
#endif


// USB Device
#include "USB/USBDesc.h"
#include "USB/USBCore.h"
#include "USB/USBAPI.h"
#include "USB/USB_host.h"

#ifdef __cplusplus
  #include "USB/CDC.h"
#endif

#endif // Arduino_h
